relengapi/blueprints/slaveloan/__init__.py

```diff
@@ -184,6 +186,53 @@ def get_machine_classes():
     which match any of the globs in the array."""
     return slave_patterns()
 
+
+@bp.route('/manual_actions')
+@p.slaveloan.admin.require()
+@apimethod([rest.ManualAction], int, bool)
+def get_loan_actions(loan_id=None, all=False):
+    "Get the manual actions for a loan"
+    action_query = ManualActions.query \
+                                .order_by(asc(ManualActions.timestamp_start))
+    if loan_id:
+        action_query = action_query.filter(ManualActions.loan_id == int(loan_id))
+    if not all:
+        action_query = action_query.filter(ManualActions.timestamp_complete.is_(None))
+    return [a.to_wsme() for a in action_query.all()]
+
+
+@bp.route('/manual_actions/<int:action_id>')
+@p.slaveloan.admin.require()
+@apimethod(rest.ManualAction, int)
+def get_loan_action(action_id):
+    "Get a specific action for a loan"
+    action = ManualActions.query.get(action_id)
+    return action.to_wsme()
+
+
+@bp.route('/manual_actions/<int:action_id>', methods=["PUT"])
+@p.slaveloan.admin.require()
+@apimethod(rest.ManualAction, int, body=rest.UpdateManualAction)
+def update_loan_action(action_id, body):
+    "Update a specific manual actions for a loan"
+    session = g.db.session('relengapi')
+    action = ManualActions.query.get(action_id)
+    if body.complete and action.timestamp_complete is None:
+        action.timestamp_complete = tz.utcnow()
+        action.complete_by = current_user.authenticated_email
+    else:
+        action.timestamp_complete = None
+        action.complete_by = None
+    session.add(action)
+    history = History(loan_id=action.loan_id,
+                      timestamp=tz.utcnow(),
+                      msg="Admin marked action (id: %s) as complete via web" %
```

Comment:
The else branch of the conditional above marks the action as *incomplete*, though.